source/wp-content/themes/wporg-showcase-2022/src/site-screenshot/view.js

```diff
@@ -2,142 +2,115 @@
+		setBase64Image( value ) {
+			const context = getContext();
+			context.base64Image = value;
+		},
 
-		if ( res.redirected ) {
-			setShouldRetry( true, store );
-		} else if ( res.status === 200 && ! res.redirected ) {
-			const blob = await res.blob();
+		*fetchImage( fullUrl ) {
+			try {
+				const res = yield fetch( fullUrl );
+				actions.increaseAttempts();
 
-			const reader = new FileReader();
-			reader.onload = ( event ) => {
-				setBase64Image( event.target.result, store );
-			};
-			reader.readAsDataURL( blob );
+				if ( res.redirected ) {
+					actions.setShouldRetry( true );
+				} else if ( res.status === 200 && ! res.redirected ) {
+					const blob = yield res.blob();
 
-			setShouldRetry( false, store );
-		}
-	} catch ( error ) {
-		setHasError( true, store );
-		setShouldRetry( false, store );
-	}
-};
+					const value = yield new Promise( ( resolve ) => {
+						const reader = new FileReader();
+						reader.onloadend = () => resolve( reader.result );
+						reader.readAsDataURL( blob );
+					} );
 
-wpStore( {
+					actions.setBase64Image( value );
+					actions.setShouldRetry( false );
+				}
+			} catch ( error ) {
+				actions.setHasError( true );
+				actions.setShouldRetry( false );
+			}
+		},
+	},
 	effects: {
-		wporg: {
-			showcase: {
-				screenshot: {
-					// Run on init, starts the image fetch process.
-					init: async ( store ) => {
-						const { context } = store;
-						const { base64Image, isMShots, src } = context.wporg.showcase.screenshot;
+		// Run on init, starts the image fetch process.
+		init: async () => {
+			const context = getContext();
+			const { base64Image, isMShots, src } = context;
+			// console.log( { base64Image, isMShots, src } );
 
-						if ( isMShots && ! base64Image ) {
-							// Initial fetch.
-							await fetchImage( src, store );
+			if ( isMShots && ! base64Image ) {
+				// Initial fetch.
+				await actions.fetchImage( src );
 
-							// Set up the function to retry every RETRY_DELAY (2 seconds).
-							const intervalId = setInterval(
-								async ( _context ) => {
-									const { attempts, base64Image: _base64Image, shouldRetry } = _context;
-									if ( shouldRetry ) {
-										await fetchImage( src, store );
-									}
-									if ( attempts >= MAX_ATTEMPTS ) {
-										clearInterval( intervalId );
-										if ( ! _base64Image ) {
-											setHasError( true, store );
-										}
-									}
-								},
-								RETRY_DELAY,
-								context.wporg.showcase.screenshot
-							);
+				// Set up the function to retry every RETRY_DELAY (2 seconds).
+				const intervalId = setInterval(
+					async ( _context ) => {
+						const { attempts, base64Image: _base64Image, shouldRetry } = _context;
+						if ( shouldRetry ) {
+							await actions.fetchImage( src );
+						}
+						if ( attempts >= MAX_ATTEMPTS ) {
+							clearInterval( intervalId );
+							if ( ! _base64Image ) {
+								actions.setHasError( true );
+							}
```

Comment:
These actions aren't working, because the callback happens outside the scope of context. Same issue happened in `fetchImage` until I refactored it to use the Promise.

<img width="432" alt="Screenshot 2024-01-25 at 6 15 59 PM" src="https://github.com/WordPress/wporg-showcase-2022/assets/541093/8c25b9f6-f633-4f66-b5ad-76b856af2c10">